apps/blockchain/lib/blockchain/ethash.ex

```diff
@@ -0,0 +1,65 @@
+defmodule Blockchain.Ethash do
+  @moduledoc """
+  This module contains the logic found in Appendix J of the
+  yellow paper concerning the Ethash implementation for POW.
+  """
+
+  @j_epoch 30_000
+  @j_datasetinit :math.pow(2, 30) |> round
+  @j_datasetgrowth :math.pow(2, 23) |> round
+  @j_mixbytes 128
+  @j_cacheinit :math.pow(2, 24) |> round
+  @j_cachegrowth :math.pow(2, 17) |> round
+  @j_hashbytes 64
+
+  @precomputed_data_sizes Path.join([__DIR__, "ethash", "data_sizes.txt"])
+                          |> File.read!()
+                          |> String.split()
+                          |> Enum.map(&String.to_integer/1)
+  @precomputed_cache_sizes Path.join([__DIR__, "ethash", "cache_sizes.txt"])
+                           |> File.read!()
+                           |> String.split()
+                           |> Enum.map(&String.to_integer/1)
+
+  def epoch(block_number) do
+    div(block_number, @j_epoch)
+  end
+
+  def dataset_size(epoch, cache \\ @precomputed_data_sizes) do
+    Enum.at(cache, epoch) || calculate_dataset_size(epoch)
+  end
+
+  defp calculate_dataset_size(epoch) do
+    largest_size(
+      @j_datasetinit + @j_datasetgrowth * epoch - @j_mixbytes,
+      factor_size: @j_mixbytes
+    )
+  end
+
+  def cache_size(epoch, cache \\ @precomputed_cache_sizes) do
+    Enum.at(cache, epoch) || calculate_cache_size(epoch)
+  end
+
+  defp calculate_cache_size(epoch) do
+    largest_size(
+      @j_cacheinit + @j_cachegrowth * epoch - @j_hashbytes,
+      factor_size: @j_hashbytes
+    )
+  end
+
+  defp largest_size(upper_bound, factor_size: unit_size) do
```

Comment:
@masonforest we went back and forth a lot on this name. How do you feel about the names now?

Reply:
I think I like `prime` over `largest_size`.

Other ideas:

`max_prime_quotient`?
`max_prime`?
`upper_bound_prime`?
`largest_size_prime`?

Reply:
The one hesitation we had with including prime in the name is that this method does not return a prime number. It returns the amount of bytes required so that when you split it up into blocks of memory (i.e. 128 or 64) you get a prime number for the amount you can store.

Some examples of what this returns include: 

```
1073739904
1082130304
1090514816
```

Which are all divisible by 128 so definitively not prime.